workspaces/pingidentity/plugins/catalog-backend-module-pingidentity/src/lib/read.ts

```diff
@@ -0,0 +1,174 @@
+  entityKind: string,
+  entityId: string
+): string => {
+  return `url:${config.apiPath}/environments/${config.envId}/${entityKind}/${entityId}`;
+}
+
+/**
+ * Returns a parsed list of all users in the Ping Identity environment
+ *
+ * @param client - The `PingIdentityClient`
+ * @param groups - A list of all groups
+ * @param groupMembersMap - Maps group ID to all user IDs that belong in that group
+ * @param userTransformer - Optional user transformer method
+ * 
+ * @returns a parsed list of all users fetched from Ping Identity of type `UserEntity`
+ */
+const parsePingIdentityUsers = async (
+  client: PingIdentityClient,
+  groups: GroupEntity[],
+  groupMembersMap: Map<string, Set<string>>,
+  userTransformer?: UserTransformer
+): Promise<UserEntity[]> => {
+  const transformer = userTransformer ?? defaultUserTransformer;
+  const users = await client.getUsers();
+  const transformedUsers: (UserEntity | undefined)[] = await Promise.all(users.map(async (user: any) => {
+    const userLocation = getEntityLocation(client.getConfig(), 'users', user.id);
+    return await transformer({
+      apiVersion: 'backstage.io/v1beta1',
+      kind: 'User',
+      metadata: {
+        name: user.username!,
+        annotations: {
+          [PING_IDENTITY_ID_ANNOTATION]: user.id!,
+          [ANNOTATION_LOCATION]: userLocation,
+          [ANNOTATION_ORIGIN_LOCATION]: userLocation,
+        },
+      },
+      spec: {
+        profile: {
+          email: user.email,
+          ...(user.name.given || user.name.family
+            ? {
+              displayName: [user.name.given, user.name.family]
+                .filter(Boolean)
+                .join(' '),
+            }
+            : {}),
+        },
+        memberOf: findGroupMemberships(user.id, groups, groupMembersMap)
+      }
+    }, client.getConfig().envId, groups);
+  }));
+  return transformedUsers.filter(user => user !== undefined) as UserEntity[];
+}
+
+/**
+ * Returns a parsed list of all groups in the Ping Identity environment
+ *
+ * @param client - The `PingIdentityClient`
+ * @param groupMembersMap - Maps group ID to all user IDs that belong in that group
+ * @param groupTransformer - Optional group transformer method
+ * 
+ * @returns a parsed list of all groups fetched from Ping Identity of type `GroupEntity`
+ */
+const parsePingIdentityGroups = async (
+  client: PingIdentityClient,
+  groupMembersMap: Map<string, Set<string>>,
+  groupTransformer?: GroupTransformer
+): Promise<GroupEntity[]> => {
+  const transformer = groupTransformer ?? defaultGroupTransformer;
+  const groups = await client.getGroups();
+  const transformedGroups: (GroupEntity | undefined)[] = await Promise.all(groups.map(async (group: any) => {
+    const groupLocation = getEntityLocation(client.getConfig(), 'groups', group.id);
+    // add users in group to group membership map
+    groupMembersMap.set(group.id, new Set(await client.getUsersInGroup(group.id)));
+    return await transformer({
```

Comment:
Same as above where I recommend also providing the response data from the API to the transformers